examples/DDIM/model.jl

```diff
@@ -0,0 +1,364 @@
+    for i in 1:(length(channels) - 2)
+        push!(down_blocks, DownBlock(channels[i], channels[i + 1], block_depth))
+    end
+    down_blocks = Lux.Chain(down_blocks...; disable_optimizations=true)
+
+    residual_blocks = []
+    push!(residual_blocks, residual_block(channels[end - 1], channels[end]))
+    for _ in 2:block_depth
+        push!(residual_blocks, residual_block(channels[end], channels[end]))
+    end
+    residual_blocks = Lux.Chain(residual_blocks...; disable_optimizations=true)
+
+    reverse!(channels)
+    up_blocks = [UpBlock(channels[i], channels[i + 1], block_depth)
+                 for i in 1:(length(channels) - 1)]
+    up_blocks = Lux.Chain(up_blocks...)
+
+    return UNet(upsample, conv_in, conv_out, down_blocks, residual_blocks, up_blocks,
+                noise_embedding)
+end
+
+function (unet::UNet)(x::Tuple{AbstractArray{T, N}, AbstractArray{T, N}}, ps,
+                      st::NamedTuple) where {T <: AbstractFloat, N}
+    noisy_images, noise_variances = x
+    @assert length(size(noisy_images)) == 4
+    @assert length(size(noise_variances)) == 4
+    @assert size(noise_variances)[1:3] == (1, 1, 1)
+    @assert size(noisy_images, 4) == size(noise_variances, 4)
+
+    emb = unet.noise_embedding(noise_variances)
+    @assert size(emb)[[1, 2, 4]] == (1, 1, size(noise_variances, 4))
+    emb, _ = unet.upsample(emb, ps.upsample, st.upsample)
+    @assert size(emb)[[1, 2, 4]] ==
+            (size(noisy_images, 1), size(noisy_images, 2), size(noise_variances, 4))
+
+    x, new_st = unet.conv_in(noisy_images, ps.conv_in, st.conv_in)
+    Lux.@set! st.conv_in = new_st
+    @assert size(x)[[1, 2, 4]] ==
+            (size(noisy_images, 1), size(noisy_images, 2), size(noisy_images, 4))
+
+    x = cat(x, emb; dims=3)
+    @assert size(x)[[1, 2, 4]] ==
+            (size(noisy_images, 1), size(noisy_images, 2), size(noisy_images, 4))
+
+    skips_at_each_stage = ()
+    for i in 1:length(unet.down_blocks)
+        layer_name = Symbol(:layer_, i)
+        (x, skips), new_st = unet.down_blocks[i](x, ps.down_blocks[layer_name],
+                                                 st.down_blocks[layer_name])
+        Lux.@set! st.down_blocks[layer_name] = new_st
+        skips_at_each_stage = (skips_at_each_stage..., skips)
+    end
+
+    x, new_st = unet.residual_blocks(x, ps.residual_blocks, st.residual_blocks)
+    Lux.@set! st.residual_blocks = new_st
+
+    for i in 1:length(unet.up_blocks)
+        layer_name = Symbol(:layer_, i)
+        x, new_st = unet.up_blocks[i]((x, skips_at_each_stage[end - i + 1]),
+                                      ps.up_blocks[layer_name], st.up_blocks[layer_name])
+        Lux.@set! st.up_blocks[layer_name] = new_st
+    end
+
+    x, new_st = unet.conv_out(x, ps.conv_out, st.conv_out)
+    Lux.@set! st.conv_out = new_st
+
+    return x, st
+end
+
+# Define DDIM model
+# This generates noise, adds it to images and calls UNet on it to denoise.
+struct DenoisingDiffusionImplicitModel <:
```

Comment:
having abstract types is not ideal. Make it `DDIM{T <: AbstractFloat}` and `(min/max)_signal_rate::T`. Ideally, all the inner fields should be typed, but that often obfuscates the error messages, so it is fine to avoid those.

Reply:
Thanks for the suggestion. I modified it.